### DISCUSSION TITLE
Add a new object within `AttributeEvents` with asyncio

### DISCUSSION THREAD

**Discussion by ehdgua01** [sqlalchemy/sqlalchemy#7346] [Usage Questions]
I'm trying to write a new history row in the `AttributeEvents` listener, but I don't know how to add a new object in the event with the async session
I got a `sqlalchemy.exc.MissingGreenlet` when I tried with `InstanceState.session`
```python
@event.listens_for(
    models.MyModel.is_active,
    "set",
    raw=True,
)
def record_new_history(instance_state, value: bool, old_value: bool, initiator):
    target = instance_state.object
    new_history = DataHistory(
        object_id=target.id,
        object_name=get_table_verbose_name(instance_state.class_),
        object_attr=target.operation.name,
        object_value=value,
        object_value_before=old_value,
        timestamp=target.updated_at,
    )
    instance_state.session.add(new_history)
    instance_state.session.flush()
...
sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/14/xd2s)
sys:1: RuntimeWarning: coroutine 'Cursor.__aenter__' was never awaited
sys:1: RuntimeWarning: coroutine 'Connection.rollback' was never awaited
```
I assumed the `InstanceState.session` is a sync session but that is not true?

**Comment by zzzeek** [comment]
instancestate.session is definitely a sync session.   the async session is there if you look at instancestate.async_session.

a complete stack trace will show more clearly what line of code is actually causing the error and why.

**Comment by ehdgua01** [comment]
Hi, 
A MissingGreenlet error occurs when trying to update the properties of the object and record it(with flushing) within the triggered event listener.

I share my code that i changed simply
```
@event.listens_for(
    models.MyModel.is_active,
    "set",
    raw=True,
)
def record_new_history(instance_state, value: bool, old_value: bool, initiator):
    target = instance_state.object
    new_history = DataHistory(
        object_id=target.id,
        object_name=get_table_verbose_name(instance_state.class_),
        object_attr=target.operation.name,
        object_value=value,
        object_value_before=old_value,
        timestamp=target.updated_at,
    )
    instance_state.session.add(new_history)
    instance_state.session.flush()  # raise MissingGreenlet exception


async def main():
    model_instance = await session.scalar(select(MyModel).where(MyModel.id == 1))
    model_instance.is_active = True  # trigger event
    await session.flush()
```

The error is disappeared when I didn't flush within the event listener

<details>
<summary>Stack trace</summary>

> Traceback (most recent call last):
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1699, in _execute_context
>     context = constructor(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 998, in _init_compiled
>     self.cursor = self.create_cursor()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1346, in create_cursor
>     return self.create_default_cursor()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1349, in create_default_cursor
>     return self._dbapi_connection.cursor()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 1083, in cursor
>     return self.dbapi_connection.cursor(*args, **kwargs)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/asyncmy.py", line 212, in cursor
>     return AsyncAdapt_asyncmy_cursor(self)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/asyncmy.py", line 58, in __init__
>     self._cursor = self.await_(cursor.__aenter__())
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 67, in await_only
>     raise exc.MissingGreenlet(
> sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/14/xd2s)
> 
> The above exception was the direct cause of the following exception:
> 
> Traceback (most recent call last):
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3440, in _flush
>     flush_context.execute()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
>     rec.execute(self)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
>     util.preloaded.orm_persistence.save_obj(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 244, in save_obj
>     _emit_insert_statements(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/persistence.py", line 1221, in _emit_insert_statements
>     result = connection._execute_20(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1611, in _execute_20
>     return meth(self, args_10style, kwargs_10style, execution_options)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/sql/elements.py", line 325, in _execute_on_connection
>     return connection._execute_clauseelement(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1478, in _execute_clauseelement
>     ret = self._execute_context(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1705, in _execute_context
>     self._handle_dbapi_exception(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2023, in _handle_dbapi_exception
>     util.raise_(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
>     raise exception
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 1699, in _execute_context
>     context = constructor(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 998, in _init_compiled
>     self.cursor = self.create_cursor()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1346, in create_cursor
>     return self.create_default_cursor()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 1349, in create_default_cursor
>     return self._dbapi_connection.cursor()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/pool/base.py", line 1083, in cursor
>     return self.dbapi_connection.cursor(*args, **kwargs)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/asyncmy.py", line 212, in cursor
>     return AsyncAdapt_asyncmy_cursor(self)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/asyncmy.py", line 58, in __init__
>     self._cursor = self.await_(cursor.__aenter__())
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 67, in await_only
>     raise exc.MissingGreenlet(
> sqlalchemy.exc.StatementError: (sqlalchemy.exc.MissingGreenlet) greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place?
> [SQL: INSERT INTO data_history (created_at, updated_at, object_id, object_name, object_attr, object_value, object_value_before, timestamp, created_by_id, updated_by_id) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)]
> [parameters: [{'object_attr': 'operation1', 'object_value_before': 'False', 'object_name': '기능별 권한', 'object_id': 130, 'object_value': 'True', 'timestamp': datetime.datetime(2021, 11, 19, 9, 4, 55, 202755), 'updated_by_id': None, 'created_by_id': None}]]
> (Background on this error at: https://sqlalche.me/e/14/xd2s)
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/Users/A202006042/Desktop/workspace/permyo/keke.py", line 84, in <module>
>     loop.run_until_complete(main())
>   File "/Users/A202006042/.pyenv/versions/3.9.4/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
>     return future.result()
>   File "/Users/A202006042/Desktop/workspace/permyo/keke.py", line 72, in main
>     await user_group_service.toggle_operations(
>   File "/Users/A202006042/Desktop/workspace/permyo/app/services/user_group.py", line 111, in toggle_operations
>     relation.is_active = operations_state[relation.operation_id]
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/attributes.py", line 459, in __set__
>     self.impl.set(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/attributes.py", line 1082, in set
>     value = self.fire_replace_event(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/attributes.py", line 1090, in fire_replace_event
>     value = fn(
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/events.py", line 2248, in wrap
>     fn(target, *arg)
>   File "/Users/A202006042/Desktop/workspace/permyo/keke.py", line 64, in record_user_group_operation_relation_history
>     instance_state.session.flush()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3340, in flush
>     self._flush(objects)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3480, in _flush
>     transaction.rollback(_capture_exception=True)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/util/langhelpers.py", line 84, in __exit__
>     compat.raise_(value, with_traceback=traceback)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
>     raise exception
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 3480, in _flush
>     transaction.rollback(_capture_exception=True)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 900, in rollback
>     util.raise_(rollback_err[1], with_traceback=rollback_err[2])
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
>     raise exception
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/orm/session.py", line 865, in rollback
>     t[1].rollback()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2337, in rollback
>     self._do_rollback()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2534, in _do_rollback
>     self._close_impl(try_deactivate=True)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2517, in _close_impl
>     self._connection_rollback_impl()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2509, in _connection_rollback_impl
>     self.connection._rollback_impl()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 965, in _rollback_impl
>     self._handle_dbapi_exception(e, None, None, None, None)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 2027, in _handle_dbapi_exception
>     util.raise_(exc_info[1], with_traceback=exc_info[2])
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
>     raise exception
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/base.py", line 963, in _rollback_impl
>     self.engine.dialect.do_rollback(self.connection)
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/engine/default.py", line 667, in do_rollback
>     dbapi_connection.rollback()
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/dialects/mysql/asyncmy.py", line 215, in rollback
>     self.await_(self._connection.rollback())
>   File "/Users/A202006042/.virtualenvs/permyo/lib/python3.9/site-packages/sqlalchemy/util/_concurrency_py3k.py", line 67, in await_only
>     raise exc.MissingGreenlet(
> sqlalchemy.exc.MissingGreenlet: greenlet_spawn has not been called; can't call await_() here. Was IO attempted in an unexpected place? (Background on this error at: https://sqlalche.me/e/14/xd2s)
> sys:1: RuntimeWarning: coroutine 'Cursor.__aenter__' was never awaited
> sys:1: RuntimeWarning: coroutine 'Connection.rollback' was never awaited
> 
> Process finished with exit code 1

</details>

**Reply by CaselIT** [reply]
Yes, I don't think it's possible to access the db inside events while using asyncio currently, since it does not run in an asyncio context (the missing greenlet error you are getting).

without looking at the code I'm not sure how feasible it is to make events run in the asyncio context. Maybe that's something we can try to address for v2? @zzzeek do you have suggestion here?

In your example the flush after the event should work though

**Reply by zzzeek** [reply]
ah right the database driver requires the greenlet context and this is an attribute event.     you should be able to use inspect(object).async_session.run_sync(lambda session: session.flush()) for this.  (Edit: right , wont work, one would need to await up at the attribute set event)

**Reply by ehdgua01** [reply]
How can use `async_session.run_sync` in synchronous function?

**Reply by CaselIT** [reply]
yes that would probably not work correctly.

I think we could open an issue to evaluate the feasibility of running events in a greenlet context

**Reply by zzzeek** [reply]
well we are talking about an attribute set:

    myobject.x = foo

if you dont have an "await" around that, there's no await to give to the database driver.

I think it's possible the notion of "calling a database function inside of an attribute set operation" where there is no explicit "await" around the outermost "set" operation is explicitly what asyncio seeks to disallow.   It's implicit IO, it's not allowed.

I think when people have these kinds of issues, it's probably good to revisit the reasons why asyncio is being used, and use that as inspiration for what the more asyncio appropriate architeture might be.

in this case I think the object needs to have a setter method that is awaitable, like:

    await myobject.set_foo_value('bar')

that seems very verbose.  but that's the idea of asyncio.   nothing gets to do IO without there being an explicit awaitable function of some kind to call.

**Reply by CaselIT** [reply]
ah I was not thinking that the source of the event was a setattr call. I guess other than calling `asyncio.run` or similar is very hard to create an async context there.
Maybe we could provide some information regarding this point in the asyncio documentation? We have a section already but I don't think this is covered

**Reply by ehdgua01** [reply]
I agree with zzzeek's comment, so I'm going to use a sync session or define a set method. thank you
And I think that adding a document for like these issues too
